src/libmongoc/tests/test-mongoc-counters.c

```diff
@@ -471,6 +471,189 @@ test_counters_streams_timeout (void)
    mongoc_client_destroy (client);
    mock_server_destroy (server);
 }
+
+static void
+test_counters_auth_with_op_msg (bool pooled)
+{
+   // This test is sensitive to the number of members in the replica set. Assert
+   // expectations to guard against the possibility of expanding the test suite
+   // to run against replica sets with a varying number of members.
+   ASSERT_WITH_MSG (test_framework_replset_member_count () == 3u,
+                    "this test requires exactly three replset members");
+
+   // SCRAM-SHA-1 is available since MongoDB server 3.0 and forces OP_MSG
+   // requests for authentication steps that follow the initial connection
+   // handshake even with speculative authentication.
+   const char *const auth_mechanism = "SCRAM-SHA-1";
+
+   // Username is also the password.
+   char *const test_user = "auth_with_op_msg";
+
+   bson_error_t error = {0};
+
+   mongoc_client_t *const setup_client = test_framework_new_default_client ();
+   mongoc_database_t *const admin =
+      mongoc_client_get_database (setup_client, "admin");
+   (void) mongoc_database_remove_user (admin, test_user, NULL);
+   ASSERT_OR_PRINT (mongoc_database_add_user (
+                       admin,
+                       test_user,
+                       test_user,
+                       tmp_bson ("{'0': {'role': 'root', 'db': 'admin'}}"),
+                       NULL,
+                       &error),
+                    error);
+
+   char *const uri_str = test_framework_get_uri_str ();
+   mongoc_uri_t *const uri = mongoc_uri_new_with_error (uri_str, &error);
+   ASSERT_OR_PRINT (uri, error);
+   mongoc_uri_set_username (uri, test_user);
+   mongoc_uri_set_password (uri, test_user);
+
+   // Specify the authentication mechanism to ensure deterministic request
+   // behavior during testing.
+   ASSERT (mongoc_uri_set_auth_mechanism (uri, auth_mechanism));
+
+   mongoc_client_pool_t *pool = NULL;
+   mongoc_client_t *client = NULL;
+
+   if (pooled) {
+      // Note: no server API version ensures OP_QUERY for initial handshake.
+      pool = test_framework_client_pool_new_from_uri (uri, NULL);
+      test_framework_set_pool_ssl_opts (pool);
+      mongoc_client_pool_set_error_api (pool, MONGOC_ERROR_API_VERSION_2);
+      client = mongoc_client_pool_pop (pool);
+   } else {
+      // Note: no server API version ensures OP_QUERY for initial handshake.
+      client = test_framework_client_new_from_uri (uri, NULL);
+      test_framework_set_ssl_opts (client);
+      mongoc_client_set_error_api (client, MONGOC_ERROR_API_VERSION_2);
+   }
+
+   mongoc_counter_auth_success_reset ();
```

Comment:
The server monitor threads are started on the first call to `mongoc_client_pool_pop`.
Suggest moving the calls to `reset` above `mongoc_client_pool_pop` to prevent a possible race between the monitoring thread sending an OP_QUERY and reseting the counter.